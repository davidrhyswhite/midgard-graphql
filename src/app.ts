import { Config } from 'apollo-server-core';
import { ApolloServer as Server } from 'apollo-server';
import { ApolloServer as Lambda } from 'apollo-server-lambda';
import { loadConfigSync } from 'graphql-config';
import schema from './schema';
import MidgardAPI from './datasources/midgard';
import { DataSources } from './types/context';

const config = loadConfigSync({ rootDir: './' });

interface ApolloConstructable<T> {
  new (args: Config): T;
}
type ServerType = Server | Lambda;

export function createServer(App: ApolloConstructable<ServerType>): ServerType {
  return new App({
    schema,
    playground: {
      config,
    },
    introspection: true,
    dataSources: (): DataSources => ({
      midgardAPI: new MidgardAPI(),
    }),
  });
}
