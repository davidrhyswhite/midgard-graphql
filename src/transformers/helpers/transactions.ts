import { assetFromString } from '@thorchain/asgardex-util';
import {
  TransactionDetail,
  Transaction,
  Sum,
  MidgardTransactionDetail,
  MidgardTx,
  MidgardSum,
} from 'midgard/transactions';

export function toSum(sum: MidgardSum): Sum {
  return {
    ...sum,
    asset: assetFromString(sum.asset),
  };
}

export function toGas(sum: MidgardSum): Sum | void {
  if (typeof sum === 'undefined') {
    return null;
  }
  return toSum(sum);
}

export function toTransaction(data: MidgardTx): Transaction {
  return {
    ...data,
    coins: data.coins.map(toSum),
  };
}

export function toTransactionsDetail(data: MidgardTransactionDetail): TransactionDetail {
  return {
    ...data,
    pool: assetFromString(data.pool),
    in: toTransaction(data.in),
    out: data.out.map(toTransaction),
    gas: toGas(data.gas),
  };
}
