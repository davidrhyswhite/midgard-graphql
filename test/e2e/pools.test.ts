import { createTestClient } from 'apollo-server-testing';
import gql from 'graphql-tag';
import nock from 'nock';
import { createPoolsDetail } from '../helpers/mock-creators';
import app from '../../src/app';

const GET_POOLS = gql`
  query getPools {
    pools {
      status
      asset {
        chain
        symbol
        ticker
      }
      price
    }
  }
`;

test('returns an AssetPool type from application', async () => {
  const { query } = createTestClient(app);

  nock('http://midgard.hostname.local').get('/pools/path').reply(200, ['AAA.BBB-CCC', 'DDD.EEE-FFF']);

  nock('http://midgard.hostname.local')
    .get('/pools/detail/path')
    .query({ asset: 'AAA.BBB-CCC' })
    .reply(200, createPoolsDetail({ asset: 'AAA.BBB-CCC', status: 'enabled' }));

  nock('http://midgard.hostname.local')
    .get('/pools/detail/path')
    .query({ asset: 'DDD.EEE-FFF' })
    .reply(200, createPoolsDetail({ asset: 'DDD.EEE-FFF', status: 'disabled' }));

  // run query against the server and snapshot the output
  const res = await query({ query: GET_POOLS });
  expect(res).toMatchSnapshot();
});
