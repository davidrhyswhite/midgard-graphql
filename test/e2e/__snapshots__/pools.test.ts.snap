// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns an AssetPool type from application 1`] = `
Object {
  "data": Object {
    "pools": Array [
      Object {
        "asset": Object {
          "chain": "AAA",
          "symbol": "BBB-CCC",
          "ticker": "BBB",
        },
        "assetDepth": "assetDepth",
        "assetROI": "assetROI",
        "assetStakedTotal": "assetStakedTotal",
        "buyAssetCount": "buyAssetCount",
        "buyFeeAverage": "buyFeeAverage",
        "buyFeesTotal": "buyFeesTotal",
        "buySlipAverage": "buySlipAverage",
        "buyTxAverage": "buyTxAverage",
        "buyVolume": "buyVolume",
        "poolDepth": "poolDepth",
        "poolFeeAverage": "poolFeeAverage",
        "poolFeesTotal": "poolFeesTotal",
        "poolROI": "poolROI",
        "poolROI12": "poolROI12",
        "poolSlipAverage": "poolSlipAverage",
        "poolStakedTotal": "poolStakedTotal",
        "poolTxAverage": "poolTxAverage",
        "poolUnits": "poolUnits",
        "poolVolume": "poolVolume",
        "poolVolume24hr": "poolVolume24hr",
        "price": "price",
        "runeDepth": "runeDepth",
        "runeROI": "runeROI",
        "runeStakedTotal": "runeStakedTotal",
        "sellAssetCount": "sellAssetCount",
        "sellFeeAverage": "sellFeeAverage",
        "sellFeesTotal": "sellFeesTotal",
        "sellSlipAverage": "sellSlipAverage",
        "sellTxAverage": "sellTxAverage",
        "sellVolume": "sellVolume",
        "stakeTxCount": "stakeTxCount",
        "stakersCount": "stakersCount",
        "stakingTxCount": "stakingTxCount",
        "status": "enabled",
        "swappersCount": "swappersCount",
        "swappingTxCount": "swappingTxCount",
        "withdrawTxCount": "withdrawTxCount",
      },
      Object {
        "asset": Object {
          "chain": "DDD",
          "symbol": "EEE-FFF",
          "ticker": "EEE",
        },
        "assetDepth": "assetDepth",
        "assetROI": "assetROI",
        "assetStakedTotal": "assetStakedTotal",
        "buyAssetCount": "buyAssetCount",
        "buyFeeAverage": "buyFeeAverage",
        "buyFeesTotal": "buyFeesTotal",
        "buySlipAverage": "buySlipAverage",
        "buyTxAverage": "buyTxAverage",
        "buyVolume": "buyVolume",
        "poolDepth": "poolDepth",
        "poolFeeAverage": "poolFeeAverage",
        "poolFeesTotal": "poolFeesTotal",
        "poolROI": "poolROI",
        "poolROI12": "poolROI12",
        "poolSlipAverage": "poolSlipAverage",
        "poolStakedTotal": "poolStakedTotal",
        "poolTxAverage": "poolTxAverage",
        "poolUnits": "poolUnits",
        "poolVolume": "poolVolume",
        "poolVolume24hr": "poolVolume24hr",
        "price": "price",
        "runeDepth": "runeDepth",
        "runeROI": "runeROI",
        "runeStakedTotal": "runeStakedTotal",
        "sellAssetCount": "sellAssetCount",
        "sellFeeAverage": "sellFeeAverage",
        "sellFeesTotal": "sellFeesTotal",
        "sellSlipAverage": "sellSlipAverage",
        "sellTxAverage": "sellTxAverage",
        "sellVolume": "sellVolume",
        "stakeTxCount": "stakeTxCount",
        "stakersCount": "stakersCount",
        "stakingTxCount": "stakingTxCount",
        "status": "disabled",
        "swappersCount": "swappersCount",
        "swappingTxCount": "swappingTxCount",
        "withdrawTxCount": "withdrawTxCount",
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
